Cython/CodeWriter.py

```diff
@@ -245,10 +249,43 @@ def visit_FuncDefNode(self, node):
         self.visit(node.body)
         self.dedent()
 
+    def visit_CFuncDefNode(self, node):
+        if node.overridable:
+            self.startline(u'cpdef ')
+        else:
+            self.startline(u'cdef ')
+        if node.modifiers:
+            self.put(' '.join(node.modifiers))
+            self.put(' ')
+        if node.visibility != 'private':
+            self.put(node.visibility)
+            self.put(u' ')
+        if node.api:
+            self.put(u'api ')
+
+        if node.base_type:
+            self.visit(node.base_type)
+            if node.base_type.name is not None:
+                self.put(u' ')
+
+        # visit the CFuncDeclaratorNode, but put a `:` at the end of line
+        self.visit(node.declarator.base)
+        self.put(u'(')
+        self.comma_separated_list(node.declarator.args)
+        self.endline(u'):')
+
+        self.indent()
+        self.visit(node.body)
+        self.dedent()
+
     def visit_CArgDeclNode(self, node):
         if node.base_type.name is not None:
             self.visit(node.base_type)
-            self.put(u" ")
+            # a special case: if `node.declarator` is a `CNameDeclaratorNode`,
+            # its "name" might be empty string, for example, for `cdef f(x)`.
+            if not isinstance(node.declarator, CNameDeclaratorNode) or \
+                    node.declarator.name:
```

Comment:
This is yet another assumption about the internal behaviour of nodes that I would prefer keeping out of the CodeWriter. Can't you just look at `node.declarator.declared_name()`?

Reply:
But I didn't find `declared_name` the this repo, any hints here?

This special case is tricky. If we don't print ` ` here and print a ` ` the visit_"declarator", there might be to spaces between type and variable.

Reply:
> `declared_name` is in master. just merge it in here.

Reply:
Fixed, thanks!

But the here the situation is still a bit complicated, since for `CSimpleBaseTypeNode`, it may have `name`, and for other types node, their `.name` is None, but do have something to print, e.g., `TemplatedTypeNode`. I have added a note for the reason in comment.